transformer_engine/common/fused_attn/fused_attn.cpp

```diff
@@ -7,16 +7,93 @@
 #include "transformer_engine/fused_attn.h"
 #include "../common.h"
 #include "utils.h"
-#include "fused_attn_fp16_bf16_max_seqlen_512.h"
+#include "fused_attn_f16_max512_seqlen.h"
+#include "fused_attn_f16_arbitrary_seqlen.h"
 #include "fused_attn_fp8.h"
 
+// select a backend for fused attention
+NVTE_Fused_Attn_Backend select_fused_attn_backend(
+        transformer_engine::DType q_dtype,
+        transformer_engine::DType kv_dtype,
+        NVTE_QKV_Layout qkv_layout,
+        NVTE_Bias_Type bias_type,
+        NVTE_Mask_Type attn_mask_type,
+        float dropout, size_t max_seqlen_q,
+        size_t max_seqlen_kv, size_t head_dim) {
+  using namespace transformer_engine;
+  NVTE_Fused_Attn_Backend backend;
+  cudaDeviceProp deviceProp;
+  int dev;
+  cudaGetDevice(&dev);
+  cudaGetDeviceProperties(&deviceProp, dev);
+  NVTE_CHECK(q_dtype == kv_dtype, "Q and KV must have the same data type.");
+  if ((q_dtype == DType::kFloat8E4M3) || (q_dtype == DType::kFloat8E5M2)
```

Comment:
We used to do this, but now the reason for changing this logic is that if there is a user input combination that fused attention doesn't support but unfused attention does, there's still a chance for users to run the attention module (might be slow) rather than erroring out in the fused attention path. This helps control the backend selection on the PyTorch level: HazyResearch Flash Attention, cuDNN fused attention, and PyTorch native unfused attention.